packages/trackerless-network/src/logic/proxy/ProxyStreamConnectionClient.ts

```diff
@@ -0,0 +1,263 @@
+    private definition?: ProxyDefinition
+    private readonly connections: Map<PeerIDKey, ProxyDirection> = new Map()
+    private readonly propagation: Propagation
+    private readonly targetNeighbors: PeerList
+    private readonly abortController: AbortController
+
+    constructor(config: ProxyStreamConnectionClientConfig) {
+        super()
+        this.config = config
+        this.rpcCommunicator = new ListeningRpcCommunicator(`layer2-${config.streamPartId}`, config.P2PTransport)
+        this.targetNeighbors = new PeerList(peerIdFromPeerDescriptor(this.config.ownPeerDescriptor), 1000)
+        this.server = new StreamNodeServer({
+            ownPeerDescriptor: this.config.ownPeerDescriptor,
+            randomGraphId: this.config.streamPartId,
+            markAndCheckDuplicate: (msg: MessageRef, prev?: MessageRef) => this.markAndCheckDuplicate(msg, prev),
+            broadcast: (message: StreamMessage, previousPeer?: string) => this.broadcast(message, previousPeer),
+            onLeaveNotice: (notice: LeaveStreamNotice) => {
+                const senderId = notice.senderId
+                const contact = this.targetNeighbors.getNeighborWithId(senderId)
+                if (contact) {
+                    setImmediate(() => this.onPeerDisconnected(contact.getPeerDescriptor()))
+                }
+            },
+            rpcCommunicator: this.rpcCommunicator
+        })
+        this.propagation = new Propagation({
+            minPropagationTargets: 2,
+            sendToNeighbor: async (neighborId: string, msg: StreamMessage): Promise<void> => {
+                const remote = this.targetNeighbors.getNeighborWithId(neighborId)
+                if (remote) {
+                    await remote.sendData(config.ownPeerDescriptor, msg)
+                } else {
+                    throw new Error('Propagation target not found')
+                }
+            }
+        })
+        this.abortController = new AbortController()
+    }
+
+    private registerDefaultServerMethods(): void {
+        this.rpcCommunicator.registerRpcNotification(StreamMessage, 'sendData',
+            (msg: StreamMessage, context) => this.server.sendData(msg, context))
+        this.rpcCommunicator.registerRpcNotification(LeaveStreamNotice, 'leaveStreamNotice',
+            (req: LeaveStreamNotice, context) => this.server.leaveStreamNotice(req, context))
+    }
+
+    private markAndCheckDuplicate(currentMessageRef: MessageRef, previousMessageRef?: MessageRef): boolean {
+        const previousNumberPair = previousMessageRef ?
+            new NumberPair(Number(previousMessageRef!.timestamp), previousMessageRef!.sequenceNumber)
+            : null
+        const currentNumberPair = new NumberPair(Number(currentMessageRef.timestamp), currentMessageRef.sequenceNumber)
+        if (!this.duplicateDetectors.has(currentMessageRef.messageChainId)) {
+            this.duplicateDetectors.set(currentMessageRef.messageChainId, new DuplicateMessageDetector(10000))
```

Comment:
The `10000` is the default in `DuplicateMessageDetector`, so we could leave it out